File: src/types/schema.ts
export interface LayoutSchema {
  id: string;
  name: string;
  version: string; // "1.0.0"
  page: PageConfig;
  fonts: FontConfig[];
  zones: {
    header: Zone;
    body: Zone;
    footer: Zone;
  };
  variables: VariableDefinition[]; // user-defined reusable values
  dataSchema: DataFieldDefinition[]; // expected input data shape
  metadata: {
    createdAt: string;
    updatedAt: string;
    author: string;
  };
}

export interface PageConfig {
  size: 'A4' | 'A5' | 'Letter' | 'Legal';
  orientation: 'portrait' | 'landscape';
  margin: {
    top: string; // CSS-like: "2cm", "20mm", "56pt"
    bottom: string;
    left: string;
    right: string;
  };
  background?: string; // hex color or image path
}

export interface FontConfig {
  family: string; // "Sarabun", "Noto Sans Thai"
  role: 'body' | 'heading' | 'mono';
  size: number; // pt
  embedded: boolean; // must be true for PDF
}

export interface Zone {
  id: string;
  components: ComponentNode[];
  minHeight?: string;
  background?: string;
  padding?: string;
  showOnFirstPageOnly?: boolean; // for header
  showOnLastPageOnly?: boolean; // for footer
}

export type ComponentNode =
  | TextComponent
  | TableComponent
  | ImageComponent
  | LineComponent
  | SpacerComponent
  | RepeaterComponent
  | ColumnLayoutComponent
  | BarcodeComponent
  | QRComponent
  | SummaryBoxComponent
  | PageBreakIndicatorComponent;

export interface BaseComponent {
  id: string;
  type: ComponentNode['type'];
  name?: string; // Custom display name for layers panel
  visible?: BindingExpression; // "{{invoice.show_discount}}"
  x?: number; // Absolute X in mm
  y?: number; // Absolute Y in mm
  width?: number; // Width in mm
  height?: number; // Height in mm
  marginTop?: string;
  marginBottom?: string;
  pageBreakBefore?: boolean;
  align?: 'left' | 'center' | 'right' | 'justify';
}

// --- Text Component ---
export interface TextComponent extends BaseComponent {
  type: 'text';
  content: string; // "{{customer.name}}" or static
  style: TextStyle;
}

export interface TextStyle {
  fontSize?: number;
  fontWeight?: 'regular' | 'medium' | 'bold';
  color?: string; // hex
  italic?: boolean;
  underline?: boolean;
  textTransform?: 'none' | 'upper' | 'lower' | 'title';
  lineHeight?: number; // scale: 1.2
  letterSpacing?: string; // e.g. "0.05em"
  justify?: boolean;
}

// --- Stroke Configuration (maps to Typst stroke dictionary) ---
export interface StrokeConfig {
  top?: string;    // e.g. "1pt + black"
  bottom?: string;
  left?: string;
  right?: string;
}

// --- Table Cell (maps to Typst table.cell) ---
export interface TableCell {
  id: string;
  content: string;        // static text or {{binding}}
  colspan?: number;
  rowspan?: number;
  align?: 'left' | 'center' | 'right';
  fill?: string;           // per-cell background color
  stroke?: StrokeConfig;   // per-cell border override
  inset?: string;          // per-cell padding override
  style?: TextStyle;       // per-cell text style
}

// --- Table Row (maps to table.header / table.footer / data rows) ---
export interface TableRow {
  id: string;
  type: 'header' | 'data' | 'footer';
  cells: TableCell[];
  height?: string;        // row height (e.g. "30pt", "auto")
  repeat?: boolean;       // for header/footer: repeat across pages
}

// --- Horizontal / Vertical Line (maps to table.hline / table.vline) ---
export interface HLineConfig {
  id: string;
  y: number;               // row position (zero-indexed)
  start?: number;          // start column (zero-indexed, inclusive)
  end?: number;            // end column (zero-indexed, exclusive)
  stroke?: string;         // e.g. "1pt + red"
  position?: 'top' | 'bottom';
}

export interface VLineConfig {
  id: string;
  x: number;               // column position (zero-indexed)
  start?: number;          // start row (zero-indexed, inclusive)
  end?: number;            // end row (zero-indexed, exclusive)
  stroke?: string;         // e.g. "1pt + blue"
  position?: 'start' | 'end';
}

// --- Fill Pattern Presets ---
export type FillPattern =
  | 'none'
  | 'striped-rows'
  | 'striped-cols'
  | 'checkerboard'
  | 'header-only'
  | 'custom';

// --- Table Component ---
export interface TableComponent extends BaseComponent {
  type: 'table';
  dataSource: BindingExpression; // "{{invoice.items}}"
  columns: TableColumn[];
  style: TableStyle;
  showHeader: boolean;
  repeatHeaderOnPage: boolean;
  isStatic?: boolean; // If true, the table does NOT loop over dataSource
  summaryRows?: SummaryRow[];
  // --- New: Structured rows for multi-row header/footer ---
  headerRows?: TableRow[];   // structured header rows
  detailRows?: TableRow[];   // structured data rows for loops (replaces strict column looping)
  footerRows?: TableRow[];   // structured footer rows
  // --- New: Manual lines ---
  hlines?: HLineConfig[];    // manual horizontal lines
  vlines?: VLineConfig[];    // manual vertical lines
}

export interface TableColumn {
  id: string;
  header: string;
  field: string; // "item.description" relative to dataSource
  width: 'auto' | '1fr' | string; // "30%", "4cm"
  align?: 'left' | 'center' | 'right';
  format?: FormatType;
  style?: TextStyle;
  borderWidth?: string;
  borderColor?: string;
  background?: string;
  colspan?: number;
  rowspan?: number;
}

export interface TableStyle {
  headerBackground?: string;
  headerTextColor?: string;
  alternateRowBackground?: string;
  borderColor?: string;
  borderWidth?: string;
  cellPadding?: string;
  fontSize?: number;
  fontWeight?: 'regular' | 'medium' | 'bold';
  lineHeight?: number;
  letterSpacing?: string;
  justify?: boolean;
  cellStyles?: Record<string, CellStyle>; // Keys: "header:col", "row:col", "data:col"
  headerRows?: number;
  footerRows?: number;
  rowHeights?: string[];
  gutter?: string; // padding between cells
  // --- New: Typst table API extensions ---
  inset?: string;                       // global cell padding (e.g. "7pt")
  fillPattern?: FillPattern;            // fill pattern preset
  stripedColor1?: string;               // even row/col color for patterns
  stripedColor2?: string;               // odd row/col color for patterns
  stroke?: string | StrokeConfig;       // global stroke config
  columnGutter?: string;                // space between columns
  rowGutter?: string;                   // space between rows
}

export interface CellStyle {
  fill?: string;
  stroke?: string | { top?: string; bottom?: string; left?: string; right?: string };
  align?: 'left' | 'center' | 'right';
  weight?: 'bold' | 'regular';
  size?: number;
  color?: string;
}

export interface SummaryRow {
  label: string;
  value: BindingExpression; // "{{formatTHB(invoice.subtotal)}}"
  style?: TextStyle;
  separator?: boolean;
}

// --- Image Component ---
export interface ImageComponent extends BaseComponent {
  type: 'image';
  src: string;       // display URL (shown in designer canvas preview)
  srcData?: string;  // base64 data URL (data:image/png;base64,...) — used by WASM
  mimeType?: string; // e.g. "image/png", "image/jpeg", "image/webp"
  width?: number;
  height?: number;
  fit?: 'cover' | 'contain' | 'stretch';
}

// --- Line Component ---
export interface LineComponent extends BaseComponent {
  type: 'line';
  style?: 'solid' | 'dashed' | 'dotted';
  color?: string;
  thickness?: string;
}

// --- Spacer Component ---
export interface SpacerComponent extends BaseComponent {
  type: 'spacer';
  height: number;
}

// --- Repeater Component ---
export interface RepeaterComponent extends BaseComponent {
  type: 'repeater';
  dataSource: BindingExpression;
  children: ComponentNode[];
  columns?: number; // multi-column layout
}

// --- Column Layout ---
export interface ColumnLayoutComponent extends BaseComponent {
  type: 'columns';
  columns: {
    width: string;
    components: ComponentNode[];
  }[];
  gap?: string;
}

// --- Barcode Component ---
export interface BarcodeComponent extends BaseComponent {
  type: 'barcode';
  value: BindingExpression;
  format: 'code128' | 'ean13' | 'pdf417';
  width?: number;
  height?: number;
}

// --- QR Component ---
export interface QRComponent extends BaseComponent {
  type: 'qr';
  value: BindingExpression;
  width?: number;
  height?: number;
}

// --- Summary Box ---
export interface SummaryBoxComponent extends BaseComponent {
  type: 'summary-box';
  rows: {
    label: string;
    value: BindingExpression;
    style?: 'normal' | 'subtotal' | 'total' | 'highlight';
  }[];
  width?: number;
}

// --- Page Break Indicator ---
export interface PageBreakIndicatorComponent extends BaseComponent {
  type: 'page-break-indicator';
  label?: string; // Optional label like "Continued on next page..."
  style?: 'solid' | 'dashed' | 'dotted';
  showPageNumber?: boolean; // Show "Page X of Y"
}

// --- Supporting Types ---
export type BindingExpression = string; // "{{...}}" or static string
export type FormatType =
  | 'text'
  | 'number'
  | 'currency-thb'
  | 'currency-usd'
  | 'date-th'
  | 'date-en'
  | 'percent'
  | 'boolean';

export interface VariableDefinition {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'color';
  defaultValue: string | number | boolean;
  description?: string;
}

export interface DataFieldDefinition {
  path: string; // "invoice.customer.name"
  type: 'string' | 'number' | 'boolean' | 'date' | 'array';
  required: boolean;
  description?: string;
  example?: unknown;
}

File: src-wasm/src/schema.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LayoutSchema {
    pub id: String,
    pub name: String,
    pub version: String,
    pub page: PageConfig,
    pub fonts: Vec<FontConfig>,
    pub zones: Zones,
    pub pages: Vec<PageDefinition>,
    #[serde(default)]
    pub groups: Vec<GroupDefinition>,
    pub group_data_source: Option<String>,
    pub batch_data_source: Option<String>,
    pub variables: Vec<VariableDefinition>,
    pub data_schema: Vec<DataFieldDefinition>,
    pub metadata: Option<serde_json::Value>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupDefinition {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub field: String,
    pub filter_by: Option<String>,
    pub sort_by: Option<String>,
    #[serde(default)]
    pub header: Option<Zone>,
    #[serde(default)]
    pub footer: Option<Zone>,
    pub repeat_header_on_page: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageDefinition {
    pub id: String,
    pub name: Option<String>,
    pub body: Zone,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageConfig {
    pub size: String,
    pub orientation: String,
    pub margin: Margin,
    pub background: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Margin {
    pub top: String,
    pub bottom: String,
    pub left: String,
    pub right: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FontConfig {
    pub family: String,
    pub role: String,
    pub size: f64,
    pub embedded: bool,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Zones {
    pub header: Zone,
    pub footer: Zone,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Zone {
    pub id: String,
    pub components: Vec<ComponentNode>,
    pub min_height: Option<String>,
    pub background: Option<String>,
    pub padding: Option<String>,
    pub show_on_first_page_only: Option<bool>,
    pub show_on_last_page_only: Option<bool>,
    pub repeat_on_every_page: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum ComponentNode {
    Text(TextComponent),
    Table(TableComponent),
    Image(ImageComponent),
    Line(LineComponent),
    Spacer(SpacerComponent),
    #[serde(rename = "summary-box")]
    SummaryBox(SummaryBoxComponent),
    Barcode(BarcodeComponent),
    Qr(QRComponent),
    Repeater(RepeaterComponent),
    Columns(ColumnsComponent),
    #[serde(rename = "page-break-indicator")]
    PageBreakIndicator(PageBreakIndicatorComponent),
    #[serde(rename = "page-number")]
    PageNumber(PageNumberComponent),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BaseComponent {
    #[serde(default)]
    pub id: String,
    pub x: Option<f64>,
    pub y: Option<f64>,
    pub width: Option<f64>,
    pub height: Option<f64>,
    pub align: Option<String>,
    pub visible: Option<String>,
    pub format: Option<String>,
    pub page_break_before: Option<bool>,
    pub margin_top: Option<f64>,
    pub margin_bottom: Option<f64>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    #[serde(default)]
    pub content: String,
    pub style: Option<TextStyle>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextStyle {
    pub font_size: Option<f64>,
    pub font_family: Option<String>,
    pub font_weight: Option<String>,
    pub color: Option<String>,
    pub italic: Option<bool>,
    pub underline: Option<bool>,
    pub line_height: Option<f64>,
    pub letter_spacing: Option<String>,
    pub justify: Option<bool>,
    pub text_transform: Option<String>,
}

// --- Individual table cell ---
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableCell {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub content: String,
    pub colspan: Option<u32>,
    pub rowspan: Option<u32>,
    pub align: Option<String>,
    pub fill: Option<String>,
    pub inset: Option<String>,
    pub stroke: Option<serde_json::Value>,
    pub style: Option<TextStyle>,
    pub format: Option<String>,
}

// --- Structured table row ---
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableRow {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub r#type: String,
    pub cells: Vec<TableCell>,
    pub height: Option<String>,
    pub repeat: Option<bool>,
}

// --- Manual horizontal line ---
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HLineConfig {
    pub id: String,
    pub y: u32,
    pub start: Option<u32>,
    pub end: Option<u32>,
    pub stroke: Option<String>,
    pub position: Option<String>,
}

// --- Manual vertical line ---
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VLineConfig {
    pub id: String,
    pub x: u32,
    pub start: Option<u32>,
    pub end: Option<u32>,
    pub stroke: Option<String>,
    pub position: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    #[serde(default)]
    pub data_source: String,
    pub columns: Vec<TableColumn>,
    pub style: Option<TableStyle>,
    pub show_header: Option<bool>,
    pub repeat_header_on_page: Option<bool>,
    pub is_static: Option<bool>,
    pub summary_rows: Option<Vec<SummaryRow>>,
    pub header_rows: Option<Vec<TableRow>>,
    pub detail_rows: Option<Vec<TableRow>>,
    pub footer_rows: Option<Vec<TableRow>>,
    pub hlines: Option<Vec<HLineConfig>>,
    pub vlines: Option<Vec<VLineConfig>>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableColumn {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub header: Option<String>,
    #[serde(default)]
    pub field: String,
    #[serde(default)]
    pub width: String,
    pub align: Option<String>,
    pub border_width: Option<String>,
    pub border_color: Option<String>,
    pub background: Option<String>,
    pub colspan: Option<u32>,
    pub rowspan: Option<u32>,
    pub format: Option<String>,
    pub style: Option<TextStyle>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableStyle {
    pub header_background: Option<String>,
    pub alternate_row_background: Option<String>,
    pub border_color: Option<String>,
    pub border_width: Option<String>,
    pub line_height: Option<f64>,
    pub letter_spacing: Option<String>,
    pub justify: Option<bool>,
    pub cell_styles: Option<HashMap<String, CellStyle>>,
    pub header_rows: Option<u32>,
    pub footer_rows: Option<u32>,
    pub row_heights: Option<Vec<String>>,
    pub inset: Option<String>,
    pub fill_pattern: Option<String>,
    pub striped_color1: Option<String>,
    pub striped_color2: Option<String>,
    pub stroke: Option<serde_json::Value>,
    pub column_gutter: Option<String>,
    pub row_gutter: Option<String>,
    pub gutter: Option<String>,
    pub font_size: Option<f64>,
    pub font_weight: Option<String>,
    pub cell_padding: Option<String>,
    pub header_text_color: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CellStyle {
    pub fill: Option<String>,
    pub stroke: Option<serde_json::Value>,
    pub align: Option<String>,
    pub weight: Option<String>,
    pub size: Option<f64>,
    pub color: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub src: String,
    pub src_data: Option<String>,
    pub mime_type: Option<String>,
    pub fit: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LineComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub color: Option<String>,
    pub thickness: Option<String>,
    pub style: Option<String>,
    pub orientation: Option<String>,
    pub cap: Option<String>,
    pub dash_array: Option<String>,
    pub stroke: Option<String>,
}


#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpacerComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub height: f64,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SummaryBoxComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub rows: Vec<SummaryRow>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SummaryRow {
    pub label: String,
    pub value: String,
    pub style: Option<String>,
    pub separator: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BarcodeComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub value: String,
    pub format: String,
    pub src: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QRComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub value: String,
    pub src: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VariableDefinition {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub r#type: String,
    #[serde(default)]
    pub default_value: serde_json::Value,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DataFieldDefinition {
    #[serde(default)]
    pub path: String,
    #[serde(default)]
    pub r#type: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RepeaterComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub data_source: String,
    pub children: Vec<ComponentNode>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnsComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub columns: Vec<ColumnDef>,
    pub gap: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnDef {
    #[serde(default)]
    pub width: String,
    pub components: Vec<ComponentNode>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageBreakIndicatorComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub label: Option<String>,
    pub style: Option<String>,
    pub show_page_number: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageNumberComponent {
    #[serde(flatten)]
    pub base: BaseComponent,
    pub format: String,
    pub style: Option<TextStyle>,
}
